README.md
Baidu_FakeS5
==============

## Version
`Rev: 1.1`

---

## Description
Baidu Fake-S5 Power Cycle Test Tool

---

## Script Process

    1. SUT: 复制SIT-Power-CycleTest-master脚本到SUT端,执行./PowerCycleTest.sh –c xx –accycle –system 
    2. Client: 复制此脚本到Client端。
    3. Client: 修改 `config.ini` 文件。 
    4. Client: 手動執行此腳本開始测试。
    5. Client: `reports` 目錄下檢查測試結果。
    6. SUT: `reports` 目錄下檢查測試結果。

## Usage

  - 可以使用 Help 信息來獲取更多有關腳本的使用方法。

    ```bash
    $ python run.py --help
    ```

  - 下面範例中會為每一個參數做解釋。

    - 以下為完整的測試命令：

      ```bash
      $ python run.py -H <BMCIP> -U <USERNAME> -P< PASSWD> power_downup -c 500 -d 180 -i 10
      ```

    - **`-H`** Remote SUT BMC IP。
    - **`-U, --username`** SUT BMC 用户名
    - **`-P, --passwd`** SUT BMC 密码
    - **`-c, --cycle`** Cycle次數。
    - **`-d, --delay`** poweroff->poweron delay时间
    - **`-i, --interval`** ping smartnic ip interval

  - 修改 `config.ini` 教學

    ```
    [HOST]
    ip = 172.17.0.62 # SUT OS IP
    osboot_timeout = 600 # 开机到进入OS超时时间
    osdelay = 480  # 进入OS后停留时间

    [CHECK_IP] # 每圈检查IP
    bf2_bmcip = 172.17.0.63
    bf2_osip = 172.17.0.62
    ```

---

## Reports

  - 可以在此目錄下 `reports/*.log` 檢查測試結果。

---

release_notes
--------------------------------------------------
Version: tv1.1
Release date: 2023-03-07
Change List:
1. modify oem command
2. modify chk_oempowerstatus function
--------------------------------------------------
Version: tv1.0
Release date: 2023-02-08
Change List: 
1. First Release
--------------------------------------------------

utils.py
import subprocess
import logging
import time
import shutil
import sys
import os
import datetime
__version__ = '0.1.0'


def clear_reports(logpath, assumeyes):
    """clear_reports(logpth, assumeyes)

    Description:
        Clear log path file

    Usage:
        clear_reports(logpath, "Parameter assumeyes")

    Parameters:
        logpath       - log path
        assumeyes - assume yes

    Return:
        None
    """
    if os.path.exists(logpath):
        if os.listdir(logpath):
            if not assumeyes:
                ans = input("WARNING: Reposts is exists, Delete all reports [y/n]")
                if ans.lower() not in ['y', 'yes']:
                    ans1 = input("Are you sure want to continue? [Y|N]:")
                    if ans1.lower() not in ['y', 'yes']:
                        exit(255)
            shutil.rmtree(logpath)
            os.makedirs(logpath)
    else:
        os.makedirs(logpath)


def exec_command(command, silent=True, ignore=False, asyncflag=False):
    """exec_command(command, silent=True, ignore=False, async=False)

    Description:
        Open a subprocess via Popen, into the OS, to execute a command

    Usage:
        result = exec_command("ls -la")

    Parameters:
        command - Command to pass to the OS as a string
        silent  - Flag to enable (False) or disable (True) logging
                  Default: True
        ignore  - Flag to log (False) or not log (True) errors (stderr)
        async   - Return subprocess Process ID

    Return:
        stdout  - Return the results of the command executed as a list
    """
    logging.debug("Run Command: " + command)
    p = subprocess.Popen(command.split(), shell=False, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    if asyncflag:
        return p.pid
    stdout, stderr = p.communicate()
    if not silent:
        logging.info(stdout)
    if p.returncode != 0 and not ignore:
        logging.error(stderr)
        raise NameError('%s failed: %s ' % (command, stderr))
    return stdout.decode()


def print_help_exit(parser, error_message):
    """print_help_exit(parser, error_message)

    Description:
        Log an error_message then exit 1

    Usage:
        print_help_exit(parser, "Parameter parser error message")

    Parameters:
        parser       - Options object
        error-mesage - Error message to log

    Return:
        Exit Code 1
    """
    logging.error(error_message)
    parser.print_help()
    sys.exit(1)


def short_ping(ip):
    """short_ping(ip)

    Description:
        Short ping test code.  Returns True if ip responds

    Usage:
        if short_ping(console_IP):

    Parameters:
        ip    - IP address to attempt to ping as a string

    Return:
        True / False
    """
    try:
        return not subprocess.call(['ping', '-c 2', ip], stdout=open(os.devnull, 'wb'))
    except subprocess.CalledProcessError as e:
        logging.error(e.output)

    return False


def ping_test(ip):
    """ping_test(ip)

    Description:
        Retry ping test code.  Returns True if ip and responds will retry 5 times before returning
        a failure response

    Usage:
        if ping_test(console_IP):

    Parameters:
        ip    - IP address to attempt to ping as a string

    Return:
        True / False
    """
    for _ in range(5):
        rcode = short_ping(ip)
        if rcode:
            return rcode
        if _ == 4 and not rcode:
            return rcode
        time.sleep(5)


def init_simple_logging(logFile):
    """init_simple_logging(logFile)

    Description:
        Initiate logging with file_name as argument. No special treatment given to
        stdout/err like the larger logging function below.

    Usage:
        init_simple_logging("/var/log/example.log")

    Parameters:
        logFile  - File name of the log file to open

    Return:
        None
    """
    logging.basicConfig(filename=logFile, level=logging.INFO, format='%(asctime)s [%(levelname)s] | %(message)s',
                        datefmt='%Y-%m-%dT%H:%M:%S')
    logging.info('Logger initialized in file %s\n' % logFile)


def initiate_logging(logFile):
    """initiate_logging(logFile)

    Description:
        Helper function to initiate logging and set up multiple logging streams.
        Allows for generation of local tests logs as well as stdout and stderr
        from parent shell

        1. The local file will collect all information from DEBUG to the standard
           INFO
        2. stdout(parent) will capture all output sent to the screen including
           ERROR
        3. stderr(parent) will only collect ERROR
        4. DEBUG is only collected in the local log.

    Usage:
        initiate_logging("/var/log/example.log")

    Parameters:
        logFile  - File name of the log file to open

    Return:
        None
    """
    format = logging.Formatter('%(asctime)s [%(levelname)s] | %(message)s', datefmt='%Y-%m-%dT%H:%M:%S')
    root = logging.getLogger()
    root.setLevel(logging.DEBUG)
    file = logging.FileHandler(logFile)
    file.setFormatter(format)
    root.addHandler(file)
    console = logging.StreamHandler(sys.stdout)
    console.setLevel(logging.DEBUG)
    console.setFormatter(format)
    root.addHandler(console)
    console2 = logging.StreamHandler()
    console2.setLevel(logging.WARNING)
    console2.setFormatter(format)
    root.addHandler(console2)
    logging.info('Logger initialized with file %s' % logFile)


def get_local_logger():
    logging.basicConfig(filename='stdout.log', format='[%(module)s %(lineno)s]%(asctime)s: %(message)s',
                        level=logging.INFO)
    return logging.getLogger('audit')


def timeout(ip, seconds, count):
    """timeout(ip, seconds, count)

    Description:
        This method will ping test IP and return if successful. If not, will
        sleep for <seconds> and retry.  If unsuccessful, will retry the
        sleep/ping cycle for <count> times

    Usage:
        result = timeout("192.168.1.1", 120, 10)

    Parameters:
        ip      - IP address to attempt ping
        seconds - Sleep time between loops
        count   - Number of loops

    Return:
        True / False
    """
    for _ in range(count):
        if ping_test(ip):
            logging.info('%s is responding to ping, continuing...' % ip)
            return True
        logging.info('%s not responding to ping, sleeping %s' % (ip, seconds))
        time.sleep(seconds)

    logging.error('%s did not come back, exiting...' % ip)
    return False


def putlog_to_file(file_name, info):
    """putlog_to_file(file_name, contents)

        Description:
            add <contents> to <file_name>
            NOTE: This is a append write

        Usage:
            putlog_to_file("/var/log/some.log", sel_dump)

        Parameters:
            file_name -- Destination file name
            contents  -- Data to be written

        Return:
            None
        """
    with open(file_name, 'a') as f:
        print(str(datetime.datetime.now()) + ': ' + str(info), file=f)


def write_to_file(file_name, contents):
    """write_to_file(file_name, contents)

    Description:
        Write <contents> to <file_name>
        NOTE: This is a destructive write

    Usage:
        write_to_file("/var/log/some.log", sel_dump)

    Parameters:
        file_name -- Destination file name
        contents  -- Data to be written

    Return:
        None
    """
    with open(file_name, 'w') as fh:
        fh.write(contents)


def read_from_file(file_name):
    """read_from_file(file_name)

    Description:
        Read <contents> from <file_name>

    Usage:
        read_from_file("/var/log/some.log")

    Parameters:
        file_name -- Source file name

    Return:
        contents  -- Data that was read
    """
    with open(file_name, 'r') as fh:
        contents = fh.read()
        return contents



run.py
import os
import sys
import logging
import time
import queue
from threading import Thread
import subprocess
from argparse import ArgumentParser
from utils import initiate_logging, short_ping, exec_command, timeout, putlog_to_file, clear_reports
import version
from lib.parserconfig import HOST_IP, OSBOOT_TIMEOUT, OSDELAY, checkip_conf
from lib.exceptions import PingIpFail, OsBootTimeOut
prog = os.path.basename(__file__)
LOGPATH = "reports"
capabilities = ["oem_power_downup", 'oem_powercycle', 'power_downup', 'powercycle']
stop_monitorflag = False
ignore_pserr = False  # Ignore oem power status code error, programs continue
ipUP = 'Up'
ipDown = 'Down'


def ping_test(check_ipdict, cycle, exp):
    logging.info("Check BF2_NIC ip status...")
    for name, ip in check_ipdict.items():
        PING_LOG = lambda msg: putlog_to_file(os.path.join(LOGPATH, ip + "_ping.log"), msg)
        PING_SUMMARY = lambda msg: putlog_to_file(os.path.join(LOGPATH, ip + "_ping_Summary.log"), msg)
        res, ret = subprocess.getstatusoutput('ping -c 2 ' + ip)
        state = ipUP if res == 0 else ipDown
        PING_LOG("Cycle# {0} >>>SUT {1} <<<\n".format(str(cycle), exp) + ret)
        # PING_LOG("\n" + ret)
        if state == exp:
            PING_SUMMARY("PASS, Cycle#{0} SUT {1} {2} {3} status is: {4}, Expect: {1}".format(cycle, exp, name, ip, state))
            logging.info("{0} {1} status is: {2}, Expect: {3} ".format(name, ip, state, exp))
        else:
            PING_SUMMARY("FAIL, Cycle#{0} SUT {1} {2} {3} status is: {4}, Expect: {1}, "
                         "Pls see {2}_ping.log for details".format(cycle, exp, name, ip, state))
            logging.error("{0} {1} status is: {2} does not match Expect: {3} , Pls see {1}_ping.log for details".format(
                name, ip, state, exp))
            raise RuntimeError('FAIL {0} {1} status does not match expectations'.format(name, ip))


def ipmonitor(ip, interval):
    global stop_monitorflag
    PING_LOG = lambda msg: putlog_to_file(os.path.join(LOGPATH, ip + "_ping.log"), msg)
    PING_SUMMARY = lambda msg: putlog_to_file(os.path.join(LOGPATH, ip + "_ping_Summary.log"), msg)
    while True:
        if stop_monitorflag == True:
            break
        ret, res = subprocess.getstatusoutput("ping -c 2 " + ip)
        PING_LOG("="*20 + "\n" + res)
        if ret:
            PING_SUMMARY("{0} not responding to ping, Pls see {0}_ping.log for details".format(ip))
            break
        else:
            PING_SUMMARY(ip + " is responding to ping, continuing...")
        time.sleep(interval)


def start_ipmonitors(checkip_conf, interval):
    logging.info("Start BF2_NIC ip monitor thread...")
    thread_list = []
    for name, ip in checkip_conf.items():
        logging.info("{0} {1} monitor thread start".format(name, ip))
        Monitor = Thread(target=ipmonitor, args=(ip, interval), name=name)
        Monitor.start()
        thread_list.append(Monitor)
    return thread_list


def chk_ipmonitor(thread_list):
    logging.info("Check BF2_NIC ipmonitor thread...")
    for thread in thread_list:
        if thread.isAlive() == True:
            logging.info("{0} {1} Monitor Thread isAlive".format(thread.name, checkip_conf[thread.name]))
        else:
            logging.error("{0} {1} Monitor Thread Dead, Pls see {1}_ping.log for details".format(
                thread.name, checkip_conf[thread.name]))
            raise RuntimeError("FAIL {0} {1} Monitor Thread Dead, Pls see {1}_ping.log for details".format(
                thread.name, checkip_conf[thread.name]))


def stop_ipmonitors(thread_list):
    global stop_monitorflag
    stop_monitorflag = True
    for thread in thread_list:
        if thread.isAlive() == True:
            thread.join()


def chk_oempowerstatus(expcode):
    logging.info("Check OEM Power status code...")
    OEM_POWERSTATE = IPMI + "raw 0x30 0x03 0x61 0x7e 0x00"
    status_dict = {"00": "S0", "04": "Fake-S5", "05": 'S5'}
    res = exec_command(OEM_POWERSTATE)
    statecode = res.split()[-1].strip()
    if statecode == expcode:
        logging.info("Current Power Status Code: {0}, Expect Code [{1}: {2}]".format(statecode, expcode, status_dict[expcode]))
    else:
        logging.error("Current Power Status Code: {0}, Expect Code [{1}: {2}]".format(statecode, expcode, status_dict[expcode]))
        if not ignore_pserr:
            raise RuntimeError("FAIL Power Status Code does not match expectations")


def wait_osboot():
    logging.info("Waitting OS Boot and IP Up...")
    if not timeout(HOST_IP, 60, OSBOOT_TIMEOUT // 60):
        raise TimeoutError("FAIL OS Boot or IP Up TimeOut {0}" + str(OSBOOT_TIMEOUT))
    logging.info("OS Boot Success")


def oem_power_downup():
    cycle = args['cycle']
    delay = args['delay']
    IPMI = "ipmitool -I lanplus -H {bmcip} -U {username} -P {passwd} ".format(**args)
    OEM_POWEROFF = IPMI + "raw 0x30 0x02 0x61 0x7e 0x00 0"
    OEM_POWERON = IPMI + "raw 0x30 0x02 0x61 0x7e 0x00 1"
    OEM_POWERSTATE = IPMI + "raw 0x30 0x03 0x61 0x7e 0x00"
    for i in range(1, cycle+1):
        logging.info("=================== Cycle# {0} ==================".format(str(i)))
        exec_command(OEM_POWEROFF, silent=False)
        time.sleep(delay)
        chk_oempowerstatus('05')
        ping_test(checkip_conf, i, ipDown)
        exec_command(OEM_POWERON)
        # logging.info("Waitting OS BOOT")
        wait_osboot()
        chk_oempowerstatus('00')
        ping_test(checkip_conf, i, ipUP)
        logging.info("Waiting {0} seconds for OS log Collection".format(str(OSDELAY)))
        time.sleep(OSDELAY)  # Wait OS Auto Collect log


def oem_powercycle():
    cycle = args['cycle']
    delay = args['delay']
    IPMI = "ipmitool -I lanplus -H {bmcip} -U {username} -P {passwd} ".format(**args)
    OEM_POWERCYCLE = IPMI + "raw 0x30 0x02 0x61 0x7e 0x00 2"
    OEM_POWERSTATE = IPMI + "raw 0x30 0x03 0x61 0x7e 0x00"
    for i in range(1, cycle+1):
        logging.info("===================Cycle# {0} ==================".format(str(i)))
        exec_command(OEM_POWERCYCLE, silent=False)
        logging.info("Sleep " + str(delay))
        time.sleep(delay)
        ping_test(checkip_conf, i, ipDown)
        wait_osboot()
        chk_oempowerstatus('00')
        ping_test(checkip_conf, i, ipUP)
        logging.info("Waiting {0} seconds for OS log Collection".format(str(OSDELAY)))
        time.sleep(OSDELAY)  # Wait OS Auto Collect log


def power_downup():
    cycle = args['cycle']
    interval = args['interval']
    delay = args['delay']
    IPMI = "ipmitool -I lanplus -H {bmcip} -U {username} -P {passwd} ".format(**args)
    POWER_OFF = IPMI + 'power off'
    POWER_ON = IPMI + 'power on'
    OEM_POWERSTATE = IPMI + "raw 0x30 0x03 0x61 0x7e 0x00"
    monitorthread_list = start_ipmonitors(checkip_conf, interval)
    try:
        for i in range(1, cycle+1):
            logging.info("===================Cycle# {0} ==================".format(str(i)))
            exec_command(POWER_OFF, silent=False)
            logging.info("Sleep " + str(delay))
            time.sleep(delay)
            chk_oempowerstatus('04')
            chk_ipmonitor(monitorthread_list)
            exec_command(POWER_ON, silent=False)
            wait_osboot()
            chk_oempowerstatus('00')
            chk_ipmonitor(monitorthread_list)
            logging.info("Waiting {0} seconds for OS log Collection".format(str(OSDELAY)))
            time.sleep(OSDELAY)  # Wait OS Auto Collect log
    #except Exception as e:
    #    logging.error(e)
    finally:
        stop_ipmonitors(monitorthread_list)


def powercycle():
    cycle = args['cycle']
    interval = args['interval']
    IPMI = "ipmitool -I lanplus -H {bmcip} -U {username} -P {passwd} ".format(**args)
    POWER_CYCLE = IPMI + 'power cycle'
    OEM_POWERSTATE = IPMI + "raw 0x30 0x03 0x61 0x7e 0x00"
    monitorthread_list = start_ipmonitors(checkip_conf, interval)
    try:
        for i in range(1, cycle+1):
            logging.info("=================== Cycle# {0} ==================".format(str(i)))
            exec_command(POWER_CYCLE, silent=False)
            time.sleep(20)
            wait_osboot()
            chk_oempowerstatus('00')
            chk_ipmonitor(monitorthread_list)
            logging.info("Waiting {0} seconds for OS log Collection".format(str(OSDELAY)))
            time.sleep(OSDELAY) # Wait OS Auto Collect log
    # except Exception as e:
    #     logging.error(e)
    finally:
        stop_ipmonitors(monitorthread_list)


def _argparse():
    parent_parser = ArgumentParser(add_help=False)
    parent_parser.add_argument("-H", metavar='', dest='bmcip', help="IPv4 address, or [IPv6 address]", required=True)
    parent_parser.add_argument("-V", '-v', action='version', version="%(prog)s version: {0}".format(version.Version),
                               help="show program's version number and exit")
    parent_parser.add_argument("-U", '--username', metavar='USERNAME', help="local username", required=True)
    parent_parser.add_argument("-P", '--passwd', metavar='PASSWORD', help="password", required=True)
    parent_parser.add_argument('-y', '--assumeyes',
                        action='store_true',
                        help='answer yes for all questions')
    parent_parser.add_argument('-D', '--debug', action='store_true', help='Enable Debug')

    parser = ArgumentParser(usage='%(prog)s [-h] [-V] [-H <HOST>] [-p <port>] -U <username> -P <password> [-j] <command> [parameter1] [parameter2]',
                                     parents=[parent_parser])

    subparsers = parser.add_subparsers()
    oem_power_downup_parser = subparsers.add_parser("oem_power_downup")
    oem_power_downup_parser.add_argument('-c', '--cycle', dest='cycle', type=int, default=500,
                                        help="Cycle Times(Default: %(default)s)")
    oem_power_downup_parser.add_argument('-d', '--delay', dest='delay', type=int, default=20,
                                        help="off to on delay time(Default: %(default)s)")
    oem_power_downup_parser.set_defaults(action=("oem_power_downup", oem_power_downup))

    oem_powercycle_parser = subparsers.add_parser("oem_powercycle")
    oem_powercycle_parser.add_argument('-c', '--cycle', dest='cycle', type=int, default=500,
                                       help="Cycle Times(Default: %(default)s)")
    oem_powercycle_parser.add_argument('-d', '--delay', dest='delay', type=int, default=20,
                                       help="down to ping BS2_BMCIP delay time(Default: %(default)s)")
    oem_powercycle_parser.set_defaults(action=("oem_powercycle", oem_powercycle))

    power_downup_parser = subparsers.add_parser("power_downup")
    power_downup_parser.add_argument('-c', '--cycle', dest='cycle', type=int, default=500,
                                    help="Cycle Times(Default: %(default)s)")
    power_downup_parser.add_argument('-d', '--delay', dest='delay', type=int, default=180,
                                    help="off to on delay time(Default: %(default)s)")
    power_downup_parser.add_argument('-i', '--interval', dest='interval', type=int, default=10,
                                     help="ping ip interval(Default: %(default)s)")
    power_downup_parser.set_defaults(action=("power_downup", power_downup))

    powercycle_parser = subparsers.add_parser("powercycle")
    powercycle_parser.add_argument('-c', '--cycle', dest='cycle', type=int, default=500, help="Cycle Times(Default: %(default)s)")
    powercycle_parser.add_argument('-i', '--interval', dest='interval', type=int, default=1,
                                     help="ping ip interval(Default: %(default)s)")
    powercycle_parser.set_defaults(action=("powercycle", powercycle))

    group1 = parser.add_argument_group('oem power cycle',
                                       'python3 %(prog)s -H <bmcip> -P <user> -U <passwd> oem_powercycle -c 2 -d 20')
    group2 = parser.add_argument_group('oem power down up',
                                       'python3 %(prog)s -H <bmcip> -P <user> -U <passwd> oem_power_downup -c 2 -d 20')
    group3 = parser.add_argument_group('powercycle',
                                       'python3 %(prog)s -H <bmcip> -P <user> -U <passwd> powercycle -c 2 -i 1')
    group4 = parser.add_argument_group('power down up',
                                       'python3 %(prog)s -H <bmcip> -P <user> -U <passwd> power_downup -c 2 -d 180 -i 10')
    return parser.parse_args().__dict__


if __name__ == '__main__':
    args = _argparse()
    #print(args)
    debug = args['debug']
    assumeyes = args['assumeyes']
    if 'action' not in args:
        print("Usage: {0} -H <BMC IP> -U <Username> -P <Password> <Command>".format(prog))
        print("Command suported list " + str(capabilities))
        print("Use -h for more help " + str(capabilities))
    else:
        IPMI = "ipmitool -I lanplus -H {bmcip} -U {username} -P {passwd} ".format(**args)
        clear_reports(LOGPATH, assumeyes)
        (name, func) = args['action']
        logFile = os.path.join(LOGPATH, "%s.log" % name)
        initiate_logging(logFile)
        func()
